### PR TITLE
from megafusion to vcf_collect

### PR DESCRIPTION
**Pull request by rannick**
Closes https://github.com/nf-core/rnafusion/issues/365 and https://github.com/nf-core/rnafusion/issues/356

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/rnafusion/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, also make a PR on the nf-core/rnafusion _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Review by peterpru — APPROVED** [review]
Nice work 💯

File: bin/vcf_collect.py [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">\n\
+##FORMAT=<ID=DV,Number=1,Type=Integer,Description="Number of paired-ends that support the event">\n\
+##FORMAT=<ID=RV,Number=1,Type=Integer,Description="Number of split reads that support the event">\n\
+##FORMAT=<ID=FFPM,Number=1,Type=Float,Description="Fusion fragments per million total RNA-seq fragments">\n\
+#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\t{}'.format(
+        sample
+    )
+
+
+def build_fusioninspector_dataframe(file):
```
Comment (fevac) [inline bin/vcf_collect.py:196]: Nice to add type hints for readability 
```suggestion
def build_fusioninspector_dataframe(file: str) -> pd.DataFrame:
```

File: bin/vcf_collect.py [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+    df = pd.read_csv(file, sep="\t")
+    df = df.rename(columns={"#FusionName": "FUSION"})
+    df[["ChromosomeA", "PosA", "Strand1"]] = df["LeftBreakpoint"].str.split(":", expand=True)
+    df[["ChromosomeB", "PosB", "Strand2"]] = df["RightBreakpoint"].str.split(":", expand=True)
+    df[["LeftGeneName", "Left_ensembl_gene_id"]] = df["LeftGene"].str.split("^", expand=True)
+    df[["RightGeneName", "Right_ensembl_gene_id"]] = df["RightGene"].str.split("^", expand=True)
+    return df.set_index(["FUSION"])
+
+
+def replace_value_with_column_name(row, value_to_replace, column_name):
```
Comment (fevac) [inline bin/vcf_collect.py:209]:
```suggestion
def replace_value_with_column_name(row: pd.Series, value_to_replace: str, column_name: str) -> str:
```

File: bin/vcf_collect.py [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+    for col_name, value in row.items():
+        if col_name == column_name:
+            if value == value_to_replace:
+                new_values = col_name
+            else:
+                new_values = ""
+    return new_values
+
+
+def concatenate_columns(row):
```
Comment (fevac) [inline bin/vcf_collect.py:223]:
```suggestion
def concatenate_columns(row: pd.Series) -> str:
```

File: bin/vcf_collect.py [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+    """
+    Concatenate non-empty values in a row into a single string separated by commas.
+    """
+    non_empty_values = [str(value) for value in row if value != ""]
+    return ",".join(non_empty_values)
+
+
+def read_build_fusionreport(fusionreport_file):
```
Comment (fevac) [inline bin/vcf_collect.py:231]:
```suggestion
def read_build_fusionreport(fusionreport_file: str) -> pd.DataFrame:
```

File: bin/vcf_collect.py [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+    fusion_report.columns = fusion_report.columns.str.upper()
+    fusion_report["FOUND_DB"] = fusion_report["FOUND_DB"].apply(lambda x: ", ".join(x))
+    fusion_report[["GeneA", "GeneB"]] = fusion_report["FUSION"].str.split("--", expand=True)
+
+    return fusion_report[["FUSION", "GeneA", "GeneB", "TOOLS_HITS", "SCORE", "FOUND_DB", "FOUND_IN"]].set_index(
+        ["FUSION"]
+    )
+
+
+def column_manipulation(df):
```
Comment (fevac) [inline bin/vcf_collect.py:269]:
```suggestion
def column_manipulation(df: pd.DataFrame) -> pd.DataFrame:
```

File: bin/vcf_collect.py [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+                row["Left_exon_number"],
+                row["Right_exon_number"],
+                row["annots"],
+            )
+        )
+        df.loc[index, "Sample"] = "./1:{}:{}:{}".format(row["JunctionReadCount"], row["SpanningFragCount"], row["FFPM"])
+    return df
+
+
+def write_vcf(df_to_print, header, out_file):
```
Comment (fevac) [inline bin/vcf_collect.py:331]:
```suggestion
def write_vcf(df_to_print: pd.DataFrame, header: str, out_file: str) -> None:
```

File: bin/vcf_collect.py [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+        index=False,
+    )
+
+    with open(out_file, "r+") as f:
+        content = f.read()
+        f.seek(0, 0)
+        f.write(header.rstrip("\r\n") + "\n" + content)
+
+
+def build_hgcn_dataframe(file):
```
Comment (fevac) [inline bin/vcf_collect.py:361]:
```suggestion
def build_hgcn_dataframe(file: str) -> pd.DataFrame:
```

File: bin/vcf_collect.py [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+    """
+    Build a DataFrame from HGNC input file, extracting 'hgnc_id' and 'ensembl_gene_id' columns.
+    """
+    df = pd.read_csv(file, sep="\t", low_memory=False)
+    df["hgnc_id"] = df["hgnc_id"].str.replace("HGNC:", "")
+    return df[["hgnc_id", "ensembl_gene_id"]].dropna()
+
+
+def build_gtf_dataframe(file):
```
Comment (fevac) [inline bin/vcf_collect.py:370]:
```suggestion
def build_gtf_dataframe(file: str) -> pd.DataFrame:
```

**Review by fevac — CHANGES_REQUESTED** [review]
Nice!! 🌟

File: bin/vcf_collect.py [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+##INFO=<ID=GENEA,Number=.,Type=String,Description="Gene A">\n\
+##INFO=<ID=GENEB,Number=.,Type=String,Description="Gene B">\n\
+##INFO=<ID=POSA,Number=.,Type=String,Description="Breakpoint position A">\n\
+##INFO=<ID=POSB,Number=.,Type=String,Description="Breakpoint position B">\n\
+##INFO=<ID=ORIENTATION,Number=.,Type=String,Description="Strand1 and strand2 directions">\n\
+##INFO=<ID=FOUND_DB,Number=.,Type=String,Description="Databases in which the fusion has been found">\n\
+##INFO=<ID=FOUND_IN,Number=.,Type=String,Description="Callers that have found the fusion">\n\
+##INFO=<ID=TOOL_HITS,Number=.,Type=Integer,Description="Number of tools that found the fusion">\n\
+##INFO=<ID=SCORE,Number=.,Type=Float,Description="Score from fusionreport">\n\
+##INFO=<ID=FRAME_STATUS,Number=.,Type=Float,Description="Frame status of the fusion">\n\
```
Comment (fevac) [inline bin/vcf_collect.py:179]:
```suggestion
##INFO=<ID=FRAME_STATUS,Number=.,Type=String,Description="Frame status of the fusion">\n\
```

File: bin/vcf_collect.py [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+##INFO=<ID=TRANSCRIPT_ID_A,Number=.,Type=Float,Description="Transcript id A ">\n\
```
Comment (fevac) [inline bin/vcf_collect.py:180]:
```suggestion
##INFO=<ID=TRANSCRIPT_ID_A,Number=.,Type=String,Description="Transcript id A ">\n\
```

File: bin/vcf_collect.py [hunk 11]
```diff
@@ -0,0 +1,373 @@
...
+##INFO=<ID=TRANSCRIPT_ID_B,Number=.,Type=Float,Description="Transcript id B">\n\
```
Comment (fevac) [inline bin/vcf_collect.py:181]:
```suggestion
##INFO=<ID=TRANSCRIPT_ID_B,Number=.,Type=String,Description="Transcript id B">\n\
```

File: bin/vcf_collect.py [hunk 12]
```diff
@@ -0,0 +1,373 @@
...
+        else:
+            df.loc[index, "ALT"] = f'N[{row["ChromosomeB"]}:{row["PosB"]}['
+
+        df.loc[index, "INFO"] = (
+            f"SVTYPE=BND;CHRA={row['ChromosomeA']};CHRB={row['ChromosomeB']};GENEA={row['GeneA']};GENEB={row['GeneB']};"
+            f"POSA={row['PosA']};POSB={row['PosB']};ORIENTATION={row['Strand1']},{row['Strand2']};FOUND_DB={row['FOUND_DB']};"
+            f"FOUND_IN={row['FOUND_IN']};TOOL_HITS={row['TOOLS_HITS']};SCORE={row['SCORE']};FRAME_STATUS={row['PROT_FUSION_TYPE']};"
+            f"TRANSCRIPT_ID_A={row['CDS_LEFT_ID']};TRANSCRIPT_ID_B={row['CDS_RIGHT_ID']};"
+            f"TRANSCRIPT_VERSION_A={row['Left_transcript_version']};TRANSCRIPT_VERSION_B={row['Right_transcript_version']};"
+            f"HGNC_ID_A={row['Left_hgnc_id']};HGNC_ID_B={row['Right_hgnc_id']};"
```
Comment (fevac) [inline bin/vcf_collect.py:306]: remove "HGNC:" prefix

File: conf/modules.config [hunk 13]
```diff
@@ -338,4 +341,9 @@ process {
             saveAs: { filename -> filename.equals('versions.yml') ? null : filename }
         ]
     }
+
+    withName: VCF_COLLECT {
+        ext.when = {!params.fusioninspector_only}
+        ext.prefix = { "${meta.id}_fusion_data" }
```
Comment (fevac) [inline conf/modules.config:347]: exclude "_fusion_data" from sample ID in the vcf header

File: bin/vcf_collect.py [hunk 14]
```diff
@@ -0,0 +1,373 @@
...
+        elif row["Strand1"] == "-" and row["Strand2"] == "-":
+            df.loc[index, "ALT"] = f'[{row["ChromosomeB"]}:{row["PosB"]}[N'
+        elif row["Strand1"] == "+" and row["Strand2"] == "-":
+            df.loc[index, "ALT"] = f'N]{row["ChromosomeB"]}:{row["PosB"]}]'
+        elif row["Strand1"] == "-" and row["Strand2"] == "+":
+            df.loc[index, "ALT"] = f'N]{row["ChromosomeB"]}:{row["PosB"]}]'
+        else:
+            df.loc[index, "ALT"] = f'N[{row["ChromosomeB"]}:{row["PosB"]}['
+
+        df.loc[index, "INFO"] = (
```
Comment (fevac) [inline bin/vcf_collect.py:300]: exclude quotes from being written in info field

**Review by fevac — COMMENTED** [review]
added the issues I found here for simplicity

**Pull request merged**